Fix/game feature

Noor-Serry opened the pull request:
# Pull Request Template

## Description
This PR addresses multiple issues and UI/UX improvements in various game screens to ensure smoother gameplay and consistent behavior.
---

## Changes Made
List the changes introduced in this pull request:

- Updated text and timer handling for the Poster game.

- Corrected border color in light theme for better visibility.

- Fixed hint functionality and removed blur effect when selecting an answer.

- Resolved duplication in question choices.

- Fixed navigation issues when pressing back in Guess Character game.

- Stopped the timer immediately when an answer is selected.

- Hid dialogs when navigating to any game screen.

- Added visual feedback when clicking hint without enough points.

- Implemented auto-navigation to next question when the timer ends.

- Ensured landscape mode support for new features.

---

## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.

https://github.com/user-attachments/assets/7ffb114a-5a60-4339-9716-2317ac3b2d0c



---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

Noor-Serry committed:
fix: correct point calculation for game answers

This commit fixes an issue where points were awarded for incorrect answers in various game types.

**Key Changes:**

- **Domain Layer:**
    - Modified `SubmitGuessMovieGenreAnswerUseCase`, `SubmitGuessReleaseYearAnswerUseCase`, `SubmitCharacterAnswerUseCase`, and `SubmitGuessMovieByPosterAnswerUseCase`:
        - Ensured that `earnedPoints` in the `AnswerResult` is set to the question's point value only if the answer is correct. Otherwise, `earnedPoints` is 0.
        - `updateUserGamePointsUseCase` (or `updatePoints`) is now only called when the answer is correct.

- **ViewModel Layer:**
    - Removed redundant checks for `answerResult.isCorrect` before adding to `totalCollectedPoints` or `totalEarnedPoints` in:
        - `GuessMovieByPosterGameViewModel`
        - `GuessGenreViewModel`
        - `GuessReleaseYearGameViewModel`
        - `GuessCharacterGameViewModel`
    - The `totalCollectedPoints` or `totalEarnedPoints` are now directly incremented by `answerResult.earnedPoints`, which will be 0 if the answer was incorrect.

Noor-Serry committed:
fix: prevent using hints after answer and improve question generation

This commit addresses two issues:

1.  **Prevent Hint Usage After Answering:**
    *   In `GuessCharacterGameViewModel`, `GuessMovieByPosterGameViewModel`, `GuessGenreViewModel`, and `GuessReleaseYearGameViewModel`, the `onHintClicked` function now checks `state.value.isNextEnabled`. If true (meaning an answer has been submitted), the hint logic is bypassed, preventing users from using hints after they've already answered.

2.  **Improved Question Generation for Character and Poster Games:**
    *   **`GenerateCharacterQuestionsUseCase`:**
        *   Now fetches `gameDifficulty.totalQuestions * 4` characters to ensure enough unique options for choices.
        *   Processes characters in chunks of 4. For each chunk, one character is selected as the correct answer, and the other three (or fewer if less than 4 are available in the chunk) become the wrong answers. This ensures varied wrong answers for each question.
    *   **`GenerateMoviePosterQuestionUseCase`:**
        *   Similarly, fetches `gameDifficulty.totalQuestions * 4` movies.
        *   Processes movies in chunks of 4, selecting one correct movie and using the others in the chunk as wrong choices.
    *   **`DoGuessMovieByPosterHintUseCase`**: Removed unused logic that modified choices, as the hint now only reduces blur.

**Other Changes:**

*   **UI Updates for Blur Effect:**
    *   `GuessCharacterGameViewModel` and `GuessMovieByPosterGameViewModel`:
        *   When a hint is used, the `blurRadius` in `QuestionUiState` is set to `5`.
        *   Added `onSubmitTheAnswerComplete` function which is called after an answer is submitted to set `blurRadius` to `0` (removing blur).
    *   `GuessCharacterUiState` and `GuessMovieByPosterUiState`: Added `blurRadius` property to `QuestionUiState` (defaulting to 8).
    *   `GuessByPosterScreen.kt` and `GuessCharacterScreen.kt`:
        *   `GuessPicture` composable now takes `blurRadius` directly from the `question.blurRadius.dp` (or `question.blurRadius` for Character).
        *   Removed local `blurRadius` state and `animateDpAsState` for blur in `GuessCharacterScreen`.
*   **Localization:**
    *   Added string resource `guess_by_poster` and updated `guess_movie_game_title` in both `values/strings.xml` and `values-ar/string.xml`.
    *   Updated `GuessByPosterScreen.kt` to use the new `guess_by_poster` string for the top bar title.

Noor-Serry committed:
refactor: stop timer on answer submission and improve navigation

This commit implements timer stopping when an answer is submitted in the "Guess Movie by Poster," "Guess Release Year," and "Guess Character" games.

It also modifies the navigation behavior so that when navigating to a game screen from the "Lets Play" tab, the "Lets Play" tab is cleared from the backstack.

Additionally, this commit:
- Removes unused imports and code.
- Comments out the launcher icon switching logic in `MainActivity` as it was causing issues.
- Removes the `NavigateToGame` effect from `GuessCharacterGameEffect` as it's no longer needed.
- Renames `onClickClose` to `onCloseButtonClicked` in `GuessCharacterInteractionListener` for clarity.

Noor-Serry committed:
refactor: update navigation logic and reset state

This commit updates the navigation behavior and state management in the `LetsPlayViewModel` and related UI components.

**Key Changes:**

- **LetsPlayViewModel:**
    - Modified `navigateToGame()` to reset `selectedGameTypeUiState`, `selectedDifficultyLevel`, and `isStartGameButtonEnable` in the UI state after navigation. This ensures a clean state when returning to the "Let's Play" screen.
- **NavigationManager:**
    - Removed unused import `com.amsterdam.viewmodel.gameEnd.ResultSideEffect`.
- **ResultScreen:**
    - Changed the navigation behavior for `ResultSideEffect.NavigateBackToMenu` from `navigationManager.toLetsPlay(clearBackStack = true)` to `navigationManager.navigateUp()`. This provides a more standard back navigation experience.

Noor-Serry committed:
refactor: Update gradient colors in LightColors

This commit updates the gradient colors for various game types in `LightColors.kt`.

- `guessCharacterStrokeGradient` now uses `Color(0x80D85895)` and `Color(0x05FFFFFF)`.
- `guessMovieByPosterStrokeGradient` now uses `Color(0x802BA3D9)` and `Color(0x05FFFFFF)`.
- `guessMovieByReleaseStrokeGradient` now uses `Color(0x800A203A)` and `Color(0x05FFFFFF)`.
- `guessMovieByGenreStrokeGradient` now uses `Color(0x80E5A02E)` and `Color(0x05FFFFFF)`.

Noor-Serry committed:
feature: display "Not Enough Points" snackbar in games

This commit implements the display of a snackbar when a user attempts an action (e.g., using a hint) but does not have enough points in the following games:

- Guess Character Game
- Guess Movie By Poster Game
- Guess Release Year Game
- Guess Genre Game

**Key Changes:**

**ViewModel Layer:**
- Added `ShowNotEnoughPointsSnackBar` effect to:
    - `GuessCharacterGameEffect.kt`
    - `GuessMovieByPosterGameEffect.kt`
    - `GuessReleaseYearGameEffect.kt`
- Modified ViewModels to send `ShowNotEnoughPointsSnackBar` effect when a `NotEnoughPointsException` is caught in the `onError` handler:
    - `GuessCharacterGameViewModel.kt`
    - `GuessMovieByPosterGameViewModel.kt`
    - `GuessReleaseYearGameViewModel.kt`
- Renamed `onError()` to `private fun onError()` and `onCompletion()` to `private fun onCompletion()` in game ViewModels.

**UI Layer:**
- Modified game screens to observe the new effect and display a snackbar using `SnackBarManager`:
    - `GuessCharacterScreen.kt`
    - `GuessByPosterScreen.kt`
    - `GuessReleaseYearGameScreen.kt`
- Updated `GuessGenreScreen.kt` to set the snackbar duration to 1500ms for consistency.

**Domain Layer:**
- Added a blank line for readability in `AflamiException.kt`.

**Navigation:**
- Removed unused `toLetsPlay` function from `NavigationManager.kt`.

Noor-Serry committed:
refactor: automatically move to next question when timer finishes

This commit modifies the behavior of the timer in several game ViewModels: `GuessMovieByPosterGameViewModel`, `GuessGenreViewModel`, `GuessReleaseYearGameViewModel`, and `GuessCharacterGameViewModel`.

Previously, when the timer finished, the UI would enable a "Next" button. Now, the game will automatically proceed to the next question or end the game if it's the last question.

Additionally, in `GuessGenreViewModel`:
- A generic `onError` function has been introduced to handle exceptions, specifically `NotEnoughPointsException` by showing a snackbar.
- The `onUseHintError` function has been removed as its functionality is now covered by the new `onError` function.

Noor-Serry committed:
feature: enhance game UI and navigation

This commit introduces several UI improvements and navigation enhancements across different game screens.

**Key Changes:**

**UI Components:**
- Added `AdaptiveAnswersColumn.kt`: A new Composable that arranges answer choices in a responsive column layout, adapting to screen width.
- Modified `GuessReleaseYearGameScreen.kt`, `GuessByPosterScreen.kt`, `GuessCharacterScreen.kt`, and `GuessGenreScreen.kt`:
    - Replaced `LazyColumn` for answer choices with the new `AdaptiveAnswersColumn`.
    - Wrapped the main content in a `Scaffold` to provide a consistent structure with a bottom bar for the "Next" button.
    - Updated navigation on game completion or cancellation to use `navigationManager.toLetsPlay(clearBackStack = true)`, ensuring the user returns to the game selection screen.
- Modified `ResultScreen.kt`:
    - Wrapped content in a `Scaffold`.
    - Made the content scrollable using `verticalScroll`.
    - Moved "Back to Menu" and "Play Again" buttons to the bottom bar of the `Scaffold`.
- Modified `LetsPlayScreen.kt`:
    - Changed `LazyColumn` to `LazyVerticalGrid` with `GridCells.Adaptive(328.dp)` to display game cards in a responsive grid.
    - Applied `navigationBarsPadding()` to the root Composable.

**Navigation:**
- Added `toLetsPlay(clearBackStack: Boolean = false)` function in `NavigationManager.kt` to navigate to the "Lets Play" tab, with an option to clear the back stack.

**ViewModel:**
- Modified `GuessGenreViewModel.kt`: Updated `onChooseAnswerClick` to directly use the answer index to fetch the selected answer from the current question's answers list.
- Modified `GenreGameInteractionListener.kt`: Changed the signature of `onChooseAnswerClick` from `(answer: MovieGenre, answerIndex: Int)` to `(answerIndex: Int)`.

Noor-Serry committed:
refactor: use AflamiException in game ViewModels

This commit updates the `onError` function in the following ViewModels to accept `AflamiException` instead of `Exception`:

- `GuessCharacterGameViewModel`
- `GuessGenreViewModel`
- `GuessMovieByPosterGameViewModel`
- `GuessReleaseYearGameViewModel`

Additionally, it removes unused imports from `ResultScreen.kt` and `LetsPlayScreen.kt`, and corrects a resource ID in `GuessGenreScreen.kt`.

mohamedd-hassan commented on viewModel/src/main/java/com/amsterdam/viewmodel/game/whichGenre/GuessGenreViewModel.kt:
```diff
@@ -187,4 +170,10 @@ class GuessGenreViewModel @Inject constructor(
+    private fun onError(error: AflamiException) {
+        when(error){
+            is NotEnoughPointsException -> sendNewEffect(GenreGameEffect.ShowNotEnoughPointsSnackBar)
+        }
```
in the design the not enough points for hint is a dialog not a snackbar, sarah created the dialog already so you can just use it

mohamedd-hassan commented on domain/src/main/java/com/amsterdam/domain/useCase/game/character/SubmitCharacterAnswerUseCase.kt:
```diff
@@ -15,12 +15,13 @@ class SubmitCharacterAnswerUseCase(
         val correct = question.correctAnswer == answer
+        var earnedPoints = 0
         if (correct) {
             updatePoints(gameDifficulty.pointsPerQuestion)
```
points should only update when the entire game is finished and not on every question solved, consider maybe using the update points usecase in the init of the result screen to update with the total earned points

Noor-Serry replied:
the both way are valid also we can discuss that in the daily meeting

Noor-Serry committed:
feat: display dialog for insufficient points in games

This commit replaces the "Not Enough Points" SnackBar with a `NotEnoughPointsDialog` in the following game screens and ViewModels:

- `GuessReleaseYearGame`
- `GuessByPosterGame`
- `GuessCharacterGame`
- `GuessGenreGame`

**Key Changes:**

**UI Layer:**
- Modified `GuessReleaseYearScreen.kt`, `GuessByPosterScreen.kt`, `GuessCharacterScreen.kt`, and `GuessGenreScreen.kt`:
    - Removed SnackBar logic for insufficient points.
    - Added `AnimatedVisibility` to display `NotEnoughPointsDialog` when `isNotEnoughPointsDialogVisible` in the UI state is true.
    - The dialog's confirm and dismiss actions are wired to a new `dismissNotEnoughPointsDialog` method in the respective interaction listeners.
- Modified `AdaptiveAnswersColumn.kt`: Removed `Modifier.width(itemWidth)` from `AnswerSelectionItem` as it seems to have been an accidental addition or no longer needed.
- Modified `NotEnoughPointsDialog.kt`: Rearranged parameters for consistency.

**ViewModel Layer:**
- For each of the affected game ViewModels (`GuessReleaseYearGameViewModel`, `GuessMovieByPosterGameViewModel`, `GuessCharacterGameViewModel`, `GuessGenreViewModel`):
    - Added `isNotEnoughPointsDialogVisible: Boolean` to their respective UI state data classes (e.g., `GuessReleaseYearUiState`).
    - Added `dismissNotEnoughPointsDialog()` to their respective interaction listener interfaces (e.g., `GuessReleaseYearInteractionListener`).
    - Implemented `dismissNotEnoughPointsDialog()` in the ViewModel to update the UI state and hide the dialog.
    - Modified the `onError` function: When a `NotEnoughPointsException` occurs, it now updates the `isNotEnoughPointsDialogVisible` state to `true` instead of sending a `ShowNotEnoughPointsSnackBar` effect.
    - Removed `ShowNotEnoughPointsSnackBar` from their respective effect sealed interfaces (e.g., `GuessReleaseYearGameEffect`).
    - When moving to the next question (either by timer finish or explicit action), `isNotEnoughPointsDialogVisible` is reset to `false`.
    - `onTimeFinish` private methods were removed, and their logic (calling `onMoveToNextQuestion`) was directly inlined in the `timerHandler.startTimer` call.

This change provides a more prominent and user-friendly way to inform the user about insufficient points for actions like using a hint.

ahmedmahgoub0 commented:
can you handle the ripple effect in `easy, medium, hard` buttons

Noor-Serry committed:
refactor(UI): make confirm button fill width in game screens

This commit updates the `ConfirmButton` in the following game screens to fill the available width and have consistent padding:
- `GuessReleaseYearGameScreen.kt`
- `GuessByPosterScreen.kt`
- `GuessCharacterScreen.kt`
- `GuessGenreScreen.kt`

Additionally, in `AdaptiveAnswersColumn.kt`, the `AnswerSelectionItem` modifier is updated to set its width to `screenWidth / itemsPerRow`.

Noor-Serry committed:
feature: add loading state to game screens

This commit introduces a loading state to the following game screens:
- `GuessReleaseYearGameScreen.kt`
- `GuessByPosterScreen.kt`
- `GuessCharacterScreen.kt`
- `GuessGenreScreen.kt`

A `LoadingContainer` is displayed with a fade-in/fade-out animation while the game questions are being fetched. The main game content is shown with a similar animation once the questions are available.

Additionally, this commit removes some `Log.e` statements from `MovieRemoteDataSourceImpl.kt`.

Noor-Serry committed:
Refactor: Remove statusBarsPadding from game screens

This commit removes the `statusBarsPadding()` modifier from the following game screens:
- `GuessReleaseYearGameScreen.kt`
- `GuessByPosterScreen.kt`
- `GuessCharacterScreen.kt`
- `ResultScreen.kt`
- `GuessGenreScreen.kt`

This change is made to ensure that the app content draws edge-to-edge, utilizing the full screen space as recommended for newer Android versions. The `navigationBarsPadding()` is kept to prevent content from being obscured by the navigation bar.

Noor-Serry committed:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into fix/game-feature

Noor-Serry commented:
@ahmedmahgoub0 I think it dose not the case here to apply ripple effect because we do not have button or component with fixed color and when the user click on it apply the ripple here we animate the color from white to pink I can increase the animate color but that will effect other part where this chip use 
<img width="1141" height="453" alt="image" src="https://github.com/user-attachments/assets/7026f6c6-c643-4e7e-8b50-1c7ac1a8eb7d" />
@ahmedmahgoub0

Noor-Serry committed:
fix: adjust ResultScreen bottom padding

This commit modifies the bottom padding of the `ResultScreen` to correctly use `innerPadding.calculateBottomPadding()`. This ensures that the screen content respects the navigation bar padding.

Previously, a fixed `16.dp` padding was applied in addition to `innerPadding`, which could lead to incorrect spacing.

Noor-Serry committed:
fix: adjust game screen padding for status bar

This commit modifies the layout of game screens to prevent content from overlapping with the system status bar.

The `LazyColumn` in `GuessReleaseYearGameScreen`, `GuessByPosterScreen`, `GuessCharacterScreen`, and `GuessGenreScreen` now applies `statusBarsPadding()` and adjusts its bottom padding using `innerPadding.calculateBottomPadding()`. This ensures the content is displayed below the status bar and above the navigation bar.

eslam-99 commented:
As @mohamedd-hassan said, calculating game point at the end of game is better as it's not appropriate using hints when you initially don't have enough points
so what happens now you start with 0pts, answered 2 easy question for 10pts, then in third question you used the 10pts you just earned, is it reasonable?

Noor-Serry commented:
I do not see any problem with that I earned point then I spent theme.
what is the problem with that also
what if the user play then he decided to out of  the game 
also we can discuss it in the daily meeting
@eslam-99

Noor-Serry commented:
1. Save points after each answer ✅

Pros:

Progress is not lost if the app crashes, the phone locks, or the user leaves the quiz midway.

You can give instant feedback and show live score updates.

Easier to sync to a server in real-time if you have online leaderboards.

Cons:

If you later allow “changing answers” before submission, you’ll have to update points (could be messy).

Slightly more storage operations (but usually negligible).

Use if:
You want real-time scoring, progress recovery, or a competitive/leaderboard feature.

2. Save points only when the quiz finishes 🏁

Pros:

No partial progress stored—cleaner “one-shot” scoring.

Easier to handle changes if the user can revisit previous questions.

Cons:

If the user exits early, all progress is lost.

Can feel less responsive (no live score updates).

Use if:
You want the user to complete the quiz in one sitting and don’t care about mid-quiz interruptions.
@eslam-99
